### PR TITLE
[None][perf] Improve PyTorch encoder-decoder support and performance

### PR DESCRIPTION
**Pull request by cascade812**
## Description

This PR broadens and optimizes encoder-decoder support in the PyTorch backend for T5, BART, and mBART. It enables overlap-scheduler execution, removes recurring decode overhead, improves T5 kernels, aligns BART and mBART generation behavior with the legacy TensorRT encoder-decoder path (which differs slightly from Hugging Face `generate()` at the output-token limit; see item 4), and adds customer-facing documentation and test coverage.

### Runtime changes

1. **Enable the overlap scheduler for encoder-decoder models** (`py_executor.py`)
   - Remove the encoder-decoder overlap-scheduler guard and run the encoder step from `_executor_loop_overlap` before decoder forward execution, so encoder-decoder models now support the PyTorch backend's default overlap scheduling.

2. **Reduce recurring decode overhead** (`model_engine.py`)
   - Reuse cross-attention metadata across stable CUDA-graph generation steps. When no new encoder request arrives and the batch composition (request IDs and cached encoder token counts) is unchanged, skip the `update_cross_metadata()` + `prepare()` calls (repeated tensor allocation and host-to-device copies) and refresh only the decoder-side references. The cached state is invalidated whenever the batch composition changes or execution leaves the CUDA-graph path.
   - Encoder-only steps now use the leaner `prepare_encoder_only()` instead of the full `prepare()`.

3. **Improve T5 kernels** (`modeling_t5.py`)
   - Thread residuals through the encoder and decoder layer stacks so the residual addition fuses into the following RMSNorm where the fused add+RMSNorm kernel is available. FP16 keeps the Hugging Face inf-clamping semantics.
   - Use the fused FlashInfer GELU-tanh-and-multiply operation for gated `gelu_new` MLPs in FP16/BF16.

### Model and generation behavior changes

4. **Seed BART/mBART forced BOS in the decoder prompt instead of a logits processor** (`llm.py`, `sampling_params.py`, `base_worker.py`, `result.py`)
   - When a BART or mBART checkpoint defines `forced_bos_token_id`, the token is now placed directly in the internal decoder prompt (`[decoder_start_token_id, forced_bos_token_id]`) rather than forced by the per-step `_BartForcedTokensLogitsProcessor`, which is removed. This eliminates per-step host work and keeps the decode path CUDA-graph and overlap friendly.
   - `SamplingParams._decoder_output_token_prefix` records the moved token; the result layer restores it to the user-visible token IDs and logprobs, and `max_tokens` deduction subtracts the prefix. The forced BOS still counts against `max_tokens`, so BART/mBART requests with forced BOS require `max_tokens >= 2` (validated with a clear error).
   - Behavior change: `forced_eos_token_id` is no longer injected when a sequence reaches `max_tokens`; the model-selected final token is preserved and `finish_reason="length"` is reported. EOS remains a natural stopping token via `end_id`. This matches the legacy TensorRT encoder-decoder path but differs from Hugging Face `generate()`, which forces EOS as the final token when a sequence reaches the output limit.

5. **Complete mBART model support** (`modeling_bart.py`)
   - Add mBART pre-norm layer ordering, final encoder/decoder stack layer norms (including weight conversion), and checkpoint-selected MLP activation via `ACT2FN`, while preserving BART post-norm behavior.

6. **Add an encoder-decoder user guide** (`docs/source/models/encoder-decoder.md`)
   - Documents supported T5, BART, and mBART checkpoints (Whisper is explicitly marked as not yet supported), the feature-support matrix, attention backend selection, KV cache managers V1/V2, greedy and beam decoding, batching, decoder CUDA graphs, overlap scheduling, tensor parallelism, serving, runtime sizing, performance, and troubleshooting.

## Performance

The following benchmarks compare the PyTorch backend with the legacy TensorRT encoder-decoder path for large-batch inference. The measurements use BF16 on one H100 80 GB GPU with greedy decoding, natural EOS stopping, an output limit of 128 tokens, and mixed encoder input lengths from 260 to 440 tokens. The Flan-T5-XL results are the average of 10 timed runs after 3 warmup runs; the BART results are the average of 20 timed runs after 5 warmup runs. Executed-token throughput includes the terminal EOS token for both paths.

### Flan-T5-XL

For Flan-T5-XL, the PyTorch backend performs on par with the legacy TensorRT path, with slightly lower latency and higher executed-token throughput across the tested batch sizes.

| Batch size | Legacy TensorRT latency | PyTorch latency | PyTorch latency improvement over legacy TensorRT | Legacy TensorRT executed tokens/s | PyTorch executed tokens/s |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 32 | 727.6 ms | 706.1 ms | 3.0% faster | 3,153 | 3,312 |
| 64 | 1,225.0 ms | 1,136.7 ms | 7.2% faster | 3,863 | 4,184 |
| 128 | 2,056.8 ms | 1,999.3 ms | 2.8% faster | 4,601 | 4,768 |

### BART-large-CNN

For BART-large-CNN, the PyTorch backend has 21.9% to 36.0% higher latency than the legacy TensorRT path across the tested batch sizes (down from 67% to 155% before the overlap scheduler and cross-attention fast path in this PR). The remaining gap is dominated by host-side input preparation and sampling; work to close it is in progress and will come in a separate PR.

| Batch size | Legacy TensorRT latency | PyTorch latency | PyTorch latency difference | Legacy TensorRT executed tokens/s | PyTorch executed tokens/s |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 32 | 229.9 ms | 280.2 ms | 21.9% slower | 12,611 | 10,662 |
| 64 | 252.2 ms | 343.0 ms | 36.0% slower | 22,007 | 16,209 |
| 128 | 352.3 ms | 472.7 ms | 34.2% slower | 31,544 | 23,518 |

## Test Coverage

- Extended the BART and T5 end-to-end tests with overlap-scheduler variants across KV cache managers V1/V2, CUDA graphs on/off, greedy decoding, and beam search; the overlap runs must produce outputs identical to the existing non-overlap runs.
- Added an mBART end-to-end test covering pre-norm model selection, source-language tokenization, forced target-language BOS, and expected greedy translation tokens.
- Test-list updates: added the overlap and mBART cases to `l0_h100.yml`; moved the single-GPU encoder-decoder suite from `l0_b200.yml` to `l0_l40s.yml`; removed the TP2 cases from `l0_dgx_b200.yml` that are already covered in `l0_dgx_h100.yml`.

## Summary by CodeRabbit

* **New Features**
  * Added support for overlap-scheduler execution in encoder-decoder generation flows.
  * Improved handling for stable generation steps to better reuse cached cross-attention state.

* **Bug Fixes**
  * Ensured logit post-processing runs at the correct time when overlap scheduling is enabled.
  * Fixed encoder-decoder overlap execution so encoder work is performed when needed.

* **Tests**
  * Expanded end-to-end coverage for encoder-decoder models to validate overlap-scheduler and non-overlap behavior produce matching results.

### CODE REVIEW

**Review by mzweilz — APPROVED** [review]
Approved for the test-db changes.

**Review by yufeiwu-nv — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]
LGTM — behavior-preserving enc-dec (T5/BART) perf work. T5 residual adds are preserved by threading (hidden_states, residual) and fusing the deferred add into the next RMSNorm; the cross-attn fast path is guarded on new_encoder_tokens==0 with identical cache/request state; decoder-only path is unaffected (forced-BOS prefix defaults empty). Covered by identical-output overlap tests.

**Review by 2ez4bz — APPROVED** [review]

File: docs/source/models/encoder-decoder.md [hunk 1]
```diff
@@ -0,0 +1,587 @@
+<!--
+SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+SPDX-License-Identifier: Apache-2.0
+-->
+
+# Use encoder-decoder models with the PyTorch backend
+
+TensorRT LLM can run supported Hugging Face encoder-decoder checkpoints directly
+with the PyTorch backend. You do not need to convert the checkpoint or build a
+TensorRT engine. The LLM API treats the supplied prompt as the encoder input and
```
Comment (Wanli-Jiang) [inline docs/source/models/encoder-decoder.md:10]: actually TRT engine path is removed recently, seems you don't need to mention that.

File: docs/source/models/encoder-decoder.md [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+| Hugging Face architecture | Model families and examples |
+| --- | --- |
+| `T5ForConditionalGeneration` | T5, Flan-T5, and ByT5, for example `google/flan-t5-small` |
+| `BartForConditionalGeneration` | BART checkpoints |
+| `MBartForConditionalGeneration` | mBART checkpoints |
+
+Whisper is not currently supported by the PyTorch encoder-decoder path. Support
+is in progress.
+
+This is the PyTorch LLM path, not the legacy TensorRT encoder-decoder workflow
```
Comment (Wanli-Jiang) [inline docs/source/models/encoder-decoder.md]: same here, no need to mention legacy trt path now.

File: tests/integration/test_lists/test-db/l0_h100.yml [hunk 3]
```diff
@@ -196,6 +196,13 @@ l0_h100:
   - llmapi/test_llm_api_pytorch_bart.py::test_bart_pytorch_generate_encoder_decoder_end_to_end[bf16-kv-v1-cuda-graph-off-beam2-bart-large-cnn]
   - llmapi/test_llm_api_pytorch_bart.py::test_bart_pytorch_generate_encoder_decoder_end_to_end[bf16-kv-v1-cuda-graph-on-beam2-bart-large-cnn]
   - llmapi/test_llm_api_pytorch_bart.py::test_bart_pytorch_generate_encoder_decoder_end_to_end[bf16-kv-v2-cuda-graph-off-greedy-bart-large-cnn]
+  - llmapi/test_llm_api_pytorch_bart.py::test_bart_pytorch_generate_encoder_decoder_end_to_end[bf16-kv-v1-cuda-graph-off-greedy-overlap-bart-large-cnn]
```
Comment (Wanli-Jiang) [inline tests/integration/test_lists/test-db/l0_h100.yml]: For long time tests, can we choose some representative tests to guard the feature? Moving the rest to post-merge stage or QA tests.

**Review by Wanli-Jiang — APPROVED** [review]
LGTM for model changes. Better to ask review from request input/output response changes, it is a bit complicated for me.

File: docs/source/models/encoder-decoder.md [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+
+
+model = "google/flan-t5-small"
+
+with LLM(
+    model=model,
+    backend="pytorch",
+    attn_backend="TRTLLM",
+    dtype="bfloat16",
+    disable_overlap_scheduler=False,
```
Comment (Tabrizian) [inline docs/source/models/encoder-decoder.md]: nit: please remove since this is the default value.

File: docs/source/models/encoder-decoder.md [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+    enable_chunked_prefill=False,
```
Comment (Tabrizian) [inline docs/source/models/encoder-decoder.md]: nit: please remove since this is the default value.

File: docs/source/models/encoder-decoder.md [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+forcing EOS. This setup also lets beam search begin after the shared decoder
+prefix without a per-step Python logits processor.
+
+### Flan-T5-XL
+
+For Flan-T5-XL, the PyTorch backend performs on par with the legacy TensorRT
+path, with slightly lower latency and higher executed-token throughput across
+the tested batch sizes.
+
+| Batch size | Legacy TensorRT latency | PyTorch latency | PyTorch latency improvement over legacy TensorRT | Legacy TensorRT executed tokens/s | PyTorch executed tokens/s |
```
Comment (Tabrizian) [inline docs/source/models/encoder-decoder.md:520]: Should we remove TRT comparison now that TRT flow is removed?

File: docs/source/models/encoder-decoder.md (hunk 1 above)
Comment (cascade812) [reply docs/source/models/encoder-decoder.md:10]: We have customers who are still using old versions to run BART/T5. We're asking to move to the pytorch path, I think it's good to keep this description here as the context.

File: tests/integration/test_lists/test-db/l0_h100.yml (hunk 3 above)
Comment (cascade812) [reply tests/integration/test_lists/test-db/l0_h100.yml]: Sure.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 7]
```diff
@@ -765,10 +765,9 @@ def __init__(
         self.inflight_req_ids = ReqIdsSet()
 
         # Encoder-decoder models execute the encoder and decoder in separate
-        # iterations. The encoder branch lives in ``_executor_loop`` only;
-        # ``_executor_loop_overlap`` has not been threaded yet. Reject
-        # pp_size > 1 for parity with the legacy TRT path (Encoder PP support
-        # is intentionally out of scope for this port).
+        # iterations in both executor loops. Reject pp_size > 1 for parity
+        # with the legacy TRT path because encoder PP send/recv support is
+        # intentionally out of scope for this port.
```
Comment (Tabrizian) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]:
```suggestion
        # TODO: Add support for pp + encoder models
```

**Review by Tabrizian — COMMENTED** [review]
Minor comments, otherwise LGTM.

File: docs/source/models/encoder-decoder.md (hunk 6 above)
Comment (cascade812) [reply docs/source/models/encoder-decoder.md:520]: For now, I’d keep it as is, since BART has not yet reached performance parity and this data helps illustrate the gap. Once parity is achieved, I’ll remove the data and update the document accordingly (will do in a separate PR)

**Review by Tabrizian — APPROVED** [review]

**Review by laikhtewari — APPROVED** [review]
